File: FE/shoea/lib/models/order_model.dart
// To parse this JSON data, do
//
//     final orderModel = orderModelFromJson(jsonString);

import 'dart:convert';

import 'package:shoea/models/ongkir/city_model.dart';

OrderModel orderModelFromJson(String str) => OrderModel.fromJson(json.decode(str));

String orderModelToJson(OrderModel data) => json.encode(data.toJson());

class OrderModel {
  OrderModel({
    this.id,
    this.transactionId,
    this.userId,
    this.shoeId,
    this.originCity,
    this.destinationCity,
    this.color,
    this.size,
    this.qty,
    this.price,
    this.payment,
    this.status,
    this.isReviewed,
    this.etd,
    this.shoe,
    this.createdAt,
  });

  int? id;
  int? transactionId;
  int? userId;
  int? shoeId;
  CityModel? originCity;
  CityModel? destinationCity;
  String? color;
  int? size;
  int? qty;
  int? price;
  String? payment;
  String? status;
  bool? isReviewed;
  String? etd;
  Shoe? shoe;
  DateTime? createdAt;

  factory OrderModel.fromJson(Map<String, dynamic> json) => OrderModel(
    id: json["id"],
    transactionId: json["transaction_id"],
    userId: json["user_id"],
    shoeId: json["shoe_id"],
    originCity: CityModel.fromJson(json["origin_city"]),
    destinationCity: CityModel.fromJson(json["destination_city"]),
    color: json["color"],
    size: json["size"],
    qty: json["qty"],
    price: json["price"],
    payment: json["payment"],
    status: json["status"],
    isReviewed: json["is_reviewed"],
    etd: json["etd"],
    shoe: Shoe.fromJson(json["shoe"]),
    createdAt: DateTime.parse(json["created_at"]),
  );

  Map<String, dynamic> toJson() => {
    "id": id,
    "transaction_id": transactionId,
    "user_id": userId,
    "shoe_id": shoeId,
    "origin_city": originCity!.toJson(),
    "destination_city": destinationCity!.toJson(),
    "color": color,
    "size": size,
    "qty": qty,
    "price": price,
    "payment": payment,
    "status": status,
    "is_reviewed": isReviewed,
    "etd": etd,
    "shoe": shoe!.toJson(),
    "created_at": createdAt!.toIso8601String(),
  };
}

class Shoe {
  Shoe({
    this.id,
    this.brand,
    this.image,
    this.title,
    this.sold,
    this.rating,
    this.review,
    this.description,
    this.sizes,
    this.colors,
    this.price,
  });

  int? id;
  String? brand;
  String? image;
  String? title;
  int? sold;
  int? rating;
  int? review;
  String? description;
  dynamic sizes;
  dynamic colors;
  int? price;

  factory Shoe.fromJson(Map<String, dynamic> json) => Shoe(
    id: json["id"],
    brand: json["brand"],
    image: json["image"],
    title: json["title"],
    sold: json["sold"],
    rating: json["rating"],
    review: json["review"],
    description: json["description"],
    sizes: json["sizes"],
    colors: json["colors"],
    price: json["price"],
  );

  Map<String, dynamic> toJson() => {
    "id": id,
    "brand": brand,
    "image": image,
    "title": title,
    "sold": sold,
    "rating": rating,
    "review": review,
    "description": description,
    "sizes": sizes,
    "colors": colors,
    "price": price,
  };
}

File: BE/shoea/models/order_model.go
package models

import (
	"fmt"
	"io/ioutil"
	"net/http"
	"shoea/db"
	"shoea/models/city"
)

func AddToOrder(userID int, shoeID int, originCityID int, destinationCityID int, price int, etd string, createdAt string) (Response, error) {
	var res Response
	var err error
	var cartID int
	var cartColor string
	var cartSize int
	var cartQty int
	var shoeSold int

	con := db.CreateCon()

	sqlStatement1 := "SELECT id, color, size, qty FROM carts WHERE user_id = ($1) AND shoe_id = ($2)"
	sqlStatement2 := "DELETE FROM carts WHERE id = ($1)"
	sqlStatement3 := "INSERT INTO orders (user_id, shoe_id, origin_city_id, destination_city_id, color, size, qty, price, status, is_reviewed, etd, created_at) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12)"
	sqlStatement4 := "SELECT sold FROM shoes WHERE id = ($1)"
	sqlStatement5 := "UPDATE shoes SET sold = ($1) WHERE id = ($2)"

	err = con.QueryRow(sqlStatement1, userID, shoeID).Scan(&cartID, &cartColor, &cartSize, &cartQty)
	if err != nil {
		return res, err
	}

	_, err = con.Exec(sqlStatement2, cartID)
	if err != nil {
		return res, err
	}

	_, err = con.Exec(sqlStatement3, userID, shoeID, originCityID, destinationCityID, cartColor, cartSize, cartQty, price, "Active", false, etd, createdAt)
	if err != nil {
		return res, err
	}

	err = con.QueryRow(sqlStatement4, shoeID).Scan(&shoeSold)
	if err != nil {
		return res, err
	}

	sold := shoeSold + cartQty
	_, err = con.Exec(sqlStatement5, sold, shoeID)
	if err != nil {
		return res, err
	}

	res.Status = http.StatusOK
	res.Message = "Success Add To Order"
	res.Data = nil

	return res, nil
}

func GetAllOrder(userID int) ([]Order, error) {
	var order Order
	arrOrder := []Order{}
	var originCityID int
	var destinationCityID int
	var shoe Shoe

	con := db.CreateCon()
	client := http.Client{}

	sqlStatement1 := "SELECT * FROM orders WHERE user_id = ($1) ORDER BY created_at DESC"
	sqlStatement2 := "SELECT brand, image, title FROM shoes WHERE id = ($1)"

	rows, err := con.Query(sqlStatement1, userID)
	if err != nil {
		return arrOrder, err
	}
	defer rows.Close()

	for rows.Next() {
		err = rows.Scan(&order.ID, &order.UserID, &order.ShoeID, &originCityID, &destinationCityID, &order.Color, &order.Size, &order.Qty, &order.Price, &order.Status, &order.IsReviewed, &order.Etd, &order.CreatedAt)
		if err != nil {
			return arrOrder, err
		}

		//origin
		urlOrigin := fmt.Sprintf("https://api.rajaongkir.com/starter/city?id=%d", originCityID)
		reqOrigin, err := http.NewRequest("GET", urlOrigin, nil)
		if err != nil {
			return arrOrder, err
		}
		reqOrigin.Header.Set("key", apiKey)

		respOrigin, err := client.Do(reqOrigin)
		if err != nil {
			return arrOrder, err
		}
		defer respOrigin.Body.Close()

		bodyBytesOrigin, err := ioutil.ReadAll(respOrigin.Body)
		if err != nil {
			return arrOrder, err
		}

		responseOrigin, err := city.UnmarshalCityModel(bodyBytesOrigin)
		if err != nil {
			return arrOrder, err
		}

		//destination
		urlDestination := fmt.Sprintf("https://api.rajaongkir.com/starter/city?id=%d", destinationCityID)
		reqDestination, err := http.NewRequest("GET", urlDestination, nil)
		if err != nil {
			return arrOrder, err
		}
		reqDestination.Header.Set("key", apiKey)

		respDestination, err := client.Do(reqDestination)
		if err != nil {
			return arrOrder, err
		}
		defer respDestination.Body.Close()

		bodyBytesDestination, err := ioutil.ReadAll(respDestination.Body)
		if err != nil {
			return arrOrder, err
		}

		responseDestination, err := city.UnmarshalCityModel(bodyBytesDestination)
		if err != nil {
			return arrOrder, err
		}

		//shoe
		err = con.QueryRow(sqlStatement2, order.ShoeID).Scan(&shoe.Brand, &shoe.Image, &shoe.Title)
		if err != nil {
			return arrOrder, err
		}

		order.OriginCity = responseOrigin.Rajaongkir.Results
		order.DestinationCity = responseDestination.Rajaongkir.Results
		order.Shoe = shoe

		arrOrder = append(arrOrder, order)
	}

	return arrOrder, nil
}
